Source/Core/Core/HW/GCMemcard/GCMemcard.cpp

```diff
@@ -1110,120 +1162,115 @@ u32 GCMemcard::ReadAnimRGBA8(u8 index, u32* buffer, u8* delays) const
   // Batten Kaitos has 0x65 as flag too. Everything but the first 3 bytes seems irrelevant.
   // Something similar happens with Wario Ware Inc. AnimSpeed
 
-  int bnrFormat = (flags & 3);
+  int bnr_format = (flags & 3);
 
-  u32 DataOffset = GetActiveDirectory().m_dir_entries[index].m_image_offset;
-  u32 DataBlock = GetActiveDirectory().m_dir_entries[index].m_first_block - MC_FST_BLOCKS;
-
-  if ((DataBlock > m_size_blocks) || (DataOffset == 0xFFFFFFFF))
-  {
+  u32 image_offset = GetActiveDirectory().m_dir_entries[index].m_image_offset;
+  if (image_offset == 0xFFFFFFFF)
     return 0;
-  }
-
-  u8* animData = (u8*)(m_data_blocks[DataBlock].m_block.data() + DataOffset);
 
-  switch (bnrFormat)
+  switch (bnr_format)
```

Comment:
I wonder if we should at least document in some way that the values 0 (which is "no banner" iirc?) and 3 (which should be CI8 with unique palette?) don't affect the `image_offset`...altho, since both 1 and 3 use a palette it feels like both might need some offsetting? Is this actually correct?

Reply:
Yeah this is sketchy, I agree. Is there any banner that uses the '3' format? Maybe we should craft our own and see how the GC BIOS interprets it...

Reply:
Okay so I played around with this in the GC BIOS and as far as I can tell:

- The first two bits of the value at 0x7 (`m_banner_and_icon_flags`) are used for the banner format.
- '0' means no banner.
- '1' means paletted (8 bit per pixel + 16 bit color palette)
- '2' means direct color (16 bit per pixel)
- '3' acts the same as '0', no banner.

So this code in `ReadAnimRGBA8()` for the offset is actually correct, strange as it seems. The code in `ReadBannerRGBA8()` is wrong, as it produces wrong results in the '3' case -- except in the case of 0xFB, which happens to give the correct result of 'no banner'.

Reply:
Okay, I updated the code with my findings and left some comments. Does this look understandable to you now?

Reply:
Seems fine! Maybe we could peek into the GC BIOS to gain more insight, but this is a good start for now.

Reply:
GC BIOS seems to agree -- in the PAL BIOS, look at 0x8131cf48 to see it mask off the last two bits and then branch depending on the four different possibilities of the value, with the branches for 0 and 3 both going to 0x8131D044.